plugin/firewall/policy/expression.go

```diff
@@ -0,0 +1,120 @@
+package policy
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/coredns/coredns/plugin/pkg/rqdata"
+
+	"github.com/coredns/coredns/request"
+
+	"github.com/coredns/coredns/plugin/metadata"
+
+	expr "github.com/Knetic/govaluate"
+)
+
+type ruleExpr struct {
+	kind        int
+	kindIfError int
+	expression  *expr.EvaluableExpression
+}
+
+// ExprEngine implement interface Engine for Firewall plugin
+// it evaluate the rues using an the lib Knetic/govaluate
+type ExprEngine struct {
+	kindIfErrorEvaluation int
+	dataFromReq           *rqdata.Mapping
+}
+
+type dataAsParam struct {
+	ctx         context.Context
+	dataFromReq *rqdata.Extractor
+}
+
+// NewExprEngine create a new Engine with default configuration
+func NewExprEngine() *ExprEngine {
+	return &ExprEngine{TypeRefuse, rqdata.NewMapping("")}
+}
+
+//BuildQueryData here return a dataAsParam that can be used by to evaluate the variables of the expression
+func (x *ExprEngine) BuildQueryData(ctx context.Context, state request.Request) (interface{}, error) {
+	return &dataAsParam{ctx, rqdata.NewExtractor(state, x.dataFromReq)}, nil
+}
+
+//BuildReplyData here return a dataAsParam that can be used by to evaluate the variables of the expression
+func (x *ExprEngine) BuildReplyData(ctx context.Context, state request.Request, query interface{}) (interface{}, error) {
+	return &dataAsParam{ctx, rqdata.NewExtractor(state, x.dataFromReq)}, nil
+}
+
+//BuildRule create a rule for Expression Engine:
+// - first param is one of the action to return
+// - second and following param is a sentence the represent an Expression
+func (x *ExprEngine) BuildRule(args []string) (Rule, error) {
+	keyword := args[0]
+	exp := args[1:]
```

Comment:
why does BuildRule get a slice of strings when the first item is clearly a keyword and the rest is an expression? Why not `keyword string, expr []string)?

Reply:
but you control the interface definition? Isn't the rule language fixed? Why do we even need that flexibility?